src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp

```diff
@@ -16,112 +16,144 @@
+        _events.ScheduleEvent(EVENT_SPELL_BANSHEECURSE, 11s);
+        _events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+        _events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+    }
 
-        void JustDied(Unit* /*killer*/) override
-        {
-            instance->SetData(TYPE_BARONESS, IN_PROGRESS);
-        }
+    void JustDied(Unit* /*killer*/) override
+    {
+        instance->SetData(TYPE_BARONESS, IN_PROGRESS);
+    }
 
-        void UpdateAI(uint32 diff) override
-        {
-            if (!UpdateVictim())
-                return;
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
 
-            //BansheeWail
-            if (BansheeWail_Timer <= diff)
-            {
-                if (rand32() % 100 < 95)
-                    DoCastVictim(SPELL_BANSHEEWAIL);
-                //4 seconds until we should cast this again
-                BansheeWail_Timer = 4000;
-            } else BansheeWail_Timer -= diff;
+        _events.Update(diff);
 
-            //BansheeCurse
-            if (BansheeCurse_Timer <= diff)
-            {
-                if (rand32() % 100 < 75)
-                    DoCastVictim(SPELL_BANSHEECURSE);
-                //18 seconds until we should cast this again
-                BansheeCurse_Timer = 18000;
-            } else BansheeCurse_Timer -= diff;
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
 
-            //Silence
-            if (Silence_Timer <= diff)
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
             {
-                if (rand32() % 100 < 80)
-                    DoCastVictim(SPELL_SILENCE);
-                //13 seconds until we should cast this again
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
-
-            DoMeleeAttackIfReady();
+            case EVENT_SPELL_BANSHEEWAIL:
+                DoCastVictim(SPELL_BANSHEEWAIL);
+                _events.ScheduleEvent(EVENT_SPELL_BANSHEEWAIL, 4s);
+                break;
+            case EVENT_SPELL_BANSHEECURSE:
+                DoCastVictim(SPELL_BANSHEECURSE);
+                _events.Repeat(18s);
+                break;
+            case EVENT_SPELL_SILENCE:
+                DoCastVictim(SPELL_SILENCE);
+                _events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+                break;
+            case EVENT_SPELL_POSSESS:
+                if (Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false))    // Random target to be possessed
+                {
+                    if (possessTarget->IsAlive())
+                    {
+                        me->CastStop();
+                        DoCast(possessTarget, SPELL_POSSESS);
+                        DoCast(possessTarget, SPELL_POSSESSED);
+                        me->SetReactState(REACT_PASSIVE);
+                        me->SetVisible(false);
```

Comment:
I really don't like this, it feels like a hack. Are you sure there's no invisibility + stun spell cast by Baroness here?

Reply:
Baroness Anastari don't have any stuns. Only the possess and silence. I was thinking about a spell, but I couldn't find any reference to that. I've found in mangos core a spell, but I'm not sure if this is the right way (because I couldn't find ANY reference).

Reply:
I suspect there's a stun+invisibility aura that she casts on herself...

Reply:
```
ID - 17250 Possess
=================================================
Category = 0, SpellIconID = 1, activeIconID = 0, SpellVisual = (4579,0)
Family SPELLFAMILY_GENERIC, flag [0] 0x00000000 [1] 0x00000000 [2] 0x00000000

SpellSchoolMask = 1 (SPELL_SCHOOL_MASK_NORMAL)
DamageClass = 0 (SPELL_DAMAGE_CLASS_NONE)
PreventionType = 0 (SPELL_PREVENTION_TYPE_NONE)
=================================================
Attributes: 0x00000190 (SPELL_ATTR0_ABILITY, SPELL_ATTR0_HIDDEN_CLIENTSIDE, SPELL_ATTR0_HIDE_IN_COMBAT_LOG)
AttributesEx1: 0x10000000 (SPELL_ATTR1_DONT_DISPLAY_IN_AURA_BAR)
=================================================
Spell Level = 0, base 0, max 0, maxTarget 0

Category = 0
DispelType = 0 (DISPEL_NONE)
Mechanic = 0 (MECHANIC_NONE)
SpellRange: (Id 1) "Self Only":
    MinRange = 0, MinRangeFriendly = 0
    MaxRange = 0, MaxRangeFriendly = 0

CastingTime (Id 1) = 0.00
Duration: ID (4)  120000, 0, 120000
Interrupt Flags: 0x00000000, AuraIF 0x00000000, ChannelIF 0x00000000
Chance = 101, charges - 0
=================================================
Effect 0: Id 6 (SPELL_EFFECT_APPLY_AURA)
BasePoints = 1
Targets (1, 0) (TARGET_UNIT_CASTER, NO_TARGET)
Aura Id 39 (SPELL_AURA_SCHOOL_IMMUNITY), value = 1, misc = 127 (127), miscB = 0, periodic = 0

Effect 1: Id 6 (SPELL_EFFECT_APPLY_AURA)
BasePoints = 5000
Targets (1, 0) (TARGET_UNIT_CASTER, NO_TARGET)
Aura Id 18 (SPELL_AURA_MOD_INVISIBILITY), value = 5000, misc = 0 (0), miscB = 0, periodic = 0

Effect 2: Id 6 (SPELL_EFFECT_APPLY_AURA)
BasePoints = 1
Targets (1, 0) (TARGET_UNIT_CASTER, NO_TARGET)
Aura Id 12 (SPELL_AURA_MOD_STUN), value = 1, misc = 0 (0), miscB = 0, periodic = 0
```

Reply:
^that's my suspicion but will check if we have sniffs

Reply:
(@Keader promised to check sniffs and see if she casts the spell, annoy him if he forgets)

Reply:
```
ServerToClient: SMSG_AURA_UPDATE (0x0496) Length: 23 ConnIdx: 0 Time: 02/19/2010 20:32:35.000 Number: 137614
GUID: Full: 0xF1300028C400C56C Type: Creature Entry: 10436 (Baroness Anastari) Low: 50540
[0] Slot: 2
[0] Spell ID: 17250 (Possess)
[0] Flags: 63 (EffectIndex0, EffectIndex1, EffectIndex2, NotCaster, Positive, Duration)
[0] Level: 59
[0] Charges: 0
[0] Max Duration: 120000
[0] Duration: 119999

ServerToClient: SMSG_SPELL_START (0x0131) Length: 31 ConnIdx: 0 Time: 02/19/2010 20:32:35.000 Number: 137615
Caster GUID: Full: 0xF1300028C400C56C Type: Creature Entry: 10436 (Baroness Anastari) Low: 50540
Caster Unit GUID: Full: 0xF1300028C400C56C Type: Creature Entry: 10436 (Baroness Anastari) Low: 50540
Cast Count: 0
Spell ID: 17250 (Possess)
Cast Flags: 11 (PendingCast, HasTrajectory, Unknown3)
Time: 0
Target Flags: 0 (Self)

ServerToClient: SMSG_SPELL_GO (0x0132) Length: 42 ConnIdx: 0 Time: 02/19/2010 20:32:35.000 Number: 137616
Caster GUID: Full: 0xF1300028C400C56C Type: Creature Entry: 10436 (Baroness Anastari) Low: 50540
Caster Unit GUID: Full: 0xF1300028C400C56C Type: Creature Entry: 10436 (Baroness Anastari) Low: 50540
Cast Count: 0
Spell ID: 17250 (Possess)
Cast Flags: 265 (PendingCast, Unknown3, Unknown7)
Time: 3880364410
Hit Count: 1
[0] Hit GUID: Full: 0xF1300028C400C56C Type: Creature Entry: 10436 (Baroness Anastari) Low: 50540
Miss Count: 0
Target Flags: 2 (Unit)
Target GUID: 0x0

[3] UpdateType: Values
[3] GUID: Full: 0xF1300028C400C56C Type: Creature Entry: 10436 (Baroness Anastari) Low: 50540
[3] UNIT_FIELD_CHARM: xxxxxx/xxxxxxx
[3] UNIT_FIELD_CHARM + 1: 41943040/1.880791E-37
[3] UNIT_FIELD_TARGET: xxxxxx/xxxxxx
[3] UNIT_FIELD_TARGET + 1: 41943040/1.880791E-37
[3] UNIT_FIELD_FLAGS: 788544/1.104985E-39


ServerToClient: SMSG_AURA_UPDATE (0x0496) Length: 30 ConnIdx: 0 Time: 02/19/2010 20:32:35.000 Number: 137627
GUID: Full: 0x28000000322F2D4 Type: Player Low: xxxxx Name: PlayerXXXX
[0] Slot: 18
[0] Spell ID: 17246 (Possessed)
[0] Flags: 161 (EffectIndex0, Duration, Negative)
[0] Level: 59
[0] Charges: 0
[0] Caster GUID: Full: 0xF1300028C400C56C Type: Creature Entry: 10436 (Baroness Anastari) Low: 50540
[0] Max Duration: 120000
[0] Duration: 120000

```

Reply:
@Treeston 17250 does the stun?

Reply:
yes, check the three effects listed in spellwork i pasted

* effect 0: immunity to all effects, applied to self
* effect 1: invisibility, strength 5000, applied to self
* effect 2: stun, applied to self